File: js/pair.mjs
function mult([b, c], [y, z]) {
  const by = b * y
  return [ by + c*y + b*z, by + c*z]
}

function ppow(pair, n) {
  if (n<2) {
    return pair
  } else {
    const n_2 = Math.floor(n/2)
    const m_2 = ppow(pair, n_2)
    let result
    // console.log(n, " m_2: ", m_2)
    const squared = mult(m_2, m_2)
    // console.log("sq: ", squared)
    if (n % 2 === 0) {
      result = squared
    } else {
      result = mult(pair, squared)
    }
    // console.log(n, " ^ ", result)
    return result
  }
}

function ipow(pair, n) {
  return ipower({frame: {p: pair, n: n, h: null, ra: 3, pc: 0}, stack:[]})
}

function ipower(state) {
  while(state)  {
    // console.log("main loop:", state)
    switch(state.frame.pc) {
      case 0: ipower0(state); break
      case 1: ipower1(state); break
      case 2: ipower2(state); break
      case 3: return state.frame.h
    }
  }
}

function ipower0(state) {
  // console.log(state.frame)
  if (state.frame.n < 2) {
    up(state, state.frame.p)
    // console.log("after up:", state)
  } else {
    state.frame.pc = 1
  }
}

function ipower1(state) {
  const n = Math.floor(state.frame.n / 2)
  state.stack.push(state.frame)
  state.frame = {n: n, p: state.frame.p, ra: 2, pc: 0}
}

function ipower2(state) {
  let sq = mult(state.frame.h, state.frame.h)
  if ( state.frame.n % 2 === 1 ) {
    sq = mult(state.frame.p, sq)
  }
  up(state, sq)
}

function up(state, value) {
   // console.log(state.stack)
  if (state.stack.length === 0) {
    state.frame.pc = 3
    state.frame.h  = value
    return
  }
  let frame = state.stack.pop()
  frame.h = value
  frame.pc = state.frame.ra
  // console.log(frame)
  state.frame = frame
}
export {ppow, ipow}

File: elixir/lib/pair.ex
defmodule Pair do

  def power(pair, 1), do: pair
  def power(pair, n) do
    p_2 = power(pair, floor(n/2))
    squared = mul(p_2, p_2)
    case rem(n, 2) do
      0 -> squared
      1 -> mul(pair, squared)
    end
    squared = mul(p_2, p_2)
    case rem(n, 2) do
      0 -> squared
      1 -> mul(pair, squared)
    end
  end

  def ipower(pair, n) when is_number(n), do: ipower({n, pair, nil}, [])
  def ipower({n, p, _}, stack) when n < 2, do: return(p, stack)
  def ipower({n, p, h}, stack) do # h should be nil here
    ipower({floor(n/2), p, nil}, [{n, p, h}|stack])
  end

  def ipower1({n, p, h}, stack) do
    squared = mul(h, h)
    case rem(n, 2) do
      0 -> return(squared, stack)
      1 -> return(mul(p, squared), stack)
    end
  end


  def mul({b, c}, {y, z}) do
    by = b*y
    {by + c*y + b*z, by + c*z}
  end

  defp return(p, []), do: p
  defp return(p, [{n, p_, _}|stack]), do: ipower1({n, p_, p},stack)
end
